grizli/prep.py

```diff
@@ -3222,6 +3270,94 @@ def process_direct_grism_visit(direct={},
+                    'F480M': [1.879639e-21, 1.453752e-30, 4.8152]}
+    if isJWST:
+        JWST_PIPELINE = os.path.dirname(jwst.__file__) + '/pipeline/'
+        steps_det1 = Detector1Pipeline.from_config_file(config_file=JWST_PIPELINE + 'calwebb_detector1.cfg')
+        for file in direct['files']:
+            hdu = pyfits.open(file, mode='update')
+            hdu[0].header['INSTRUME'] = 'WFC3'
+            hdu[0].header['DETECTOR'] = 'IR'
+            hdu[1].header['NGOODPIX'] = -99
+            hdu[1].header['EXPNAME'] = hdu[0].header['EXPOSURE']
+            hdu[1].header['MEANDARK'] = -99
+            hdu[1].header['IDCSCALE'] = 0.065 ## how can I get this automatically?
+            hdu[0].header['PHOTFLAM'] = phot_keywords[hdu[0].header['PUPIL']][0]
+            hdu[0].header['PHOTFNU'] = phot_keywords[hdu[0].header['PUPIL']][1]
+            hdu[0].header['PHOTPLAM'] = phot_keywords[hdu[0].header['PUPIL']][2] * 10000 # microns to angstroms
+            gain_file = steps_det1.gain_scale.get_reference_file(file, 'gain')
+            gain_im = pyfits.open(gain_file)
+            im = pyfits.open(file)
+            gain_median = np.median(gain_im[1].data)
+            im['SCI'].data *= gain_median #gain_im['SCI'].data
+            im['SCI'].header['BUNIT'] = 'ELECTRONS/s'
+            im['ERR'].data *= gain_median #gain_im['SCI'].data
+            im['ERR'].header['BUNIT'] = 'ELECTRONS/s'
+            gain_im.close()
+            im.close()
+            if hdu[0].header['PUPIL'] == 'F115W':
+                hdu[0].header['PFLTFILE'] = os.path.join(os.getenv('jref'), 'jwst_niriss_flat_0193.fits')
+            if hdu[0].header['PUPIL'] == 'F150W':
+                hdu[0].header['PFLTFILE'] = os.path.join(os.getenv('jref'), 'jwst_niriss_flat_0007.fits')
+            if hdu[0].header['PUPIL'] == 'F200W':
+                hdu[0].header['PFLTFILE'] = os.path.join(os.getenv('jref'), 'jwst_niriss_flat_0189.fits')
+            
+            hdu.flush()
+    #
+        for file in grism['files']:
+            hdu = pyfits.open(file, mode='update')
+            hdu[0].header['INSTRUME'] = 'WFC3'
+            hdu[0].header['DETECTOR'] = 'IR'
+            hdu[1].header['NGOODPIX'] = -99
+            hdu[1].header['EXPNAME'] = hdu[0].header['EXPOSURE']
+            hdu[1].header['MEANDARK'] = -99
+            hdu[1].header['IDCSCALE'] = 0.065 ## get this automatically when pixelscale is in image headers
+            hdu[0].header['PHOTFLAM'] = phot_keywords[hdu[0].header['PUPIL']][0]
+            hdu[0].header['PHOTFNU'] = phot_keywords[hdu[0].header['PUPIL']][1]
+            hdu[0].header['PHOTPLAM'] = phot_keywords[hdu[0].header['PUPIL']][2] * 10000 # microns to angstroms
+            gain_file = steps_det1.gain_scale.get_reference_file(file, 'gain')
+            gain_im = pyfits.open(gain_file)
+            if (hdu[0].header['FILTER'] == 'GR150C') & (hdu[0].header['PUPIL'] == 'F115W'):        
```

Comment:
See a comment above for specifying the flat list as a dictionary, which will need to be specified in just one place for use here and also where it came up above.  Maybe a more elegant solution would be having an explicit separate function that returns all necessary keywords based on an input header, like 

```python
_PHOTKEYS_YML = """
PHOTFLAM: 
  NIRISS:
    F090W: 1.098934e-20
    F115W: 1.23456
  NIRCAM:
    F090W: 3.4567

PHOTFNU:
  NIRISS:
    F090W: 9.99999
    F115W: 9.99999
  NIRCAM:
    F090W: 9.999
        
PHOTPLAM:
  NIRISS:
    F090W: 9.99999
    F115W: 9.99999
  NIRCAM:
    F090W: 9.999
"""
_PHOTKEYS = yaml.safe_load(_PHOTKEYS_YML)

def get_hst_photom_keywords(hdu):
    get PHOTKEYS based on hdu[0].header
```

And the HDU updates could be done in separate functions that would be easier to maintain and extend, e.g., in pseudocode

```python
def jwst_hdu_like_hst(hdu):
    if niriss:
        niriss_hdu_like_hst(hdu)
    elif nircam:
        nircam_hdu_like_hst(hdu)
    elif miri:
        miri_hdu_like_hst(hdu)

def niriss_hdu_like_hst(hdu):
    # ... set keywords etc.
   
...
```
And if the gain correction needs to be applied,  it should be done above where you do the flat correction.  And better still, the flat and gain corrections should be done by the `jwst` pipeline steps themselves.